### PR TITLE
[SPARK-8964] [SQL] Use Exchange to perform shuffle in Limit

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch changes the implementation of the physical `Limit` operator so that it relies on the `Exchange` operator to perform data movement rather than directly using `ShuffledRDD`. In addition to improving efficiency, this lays the necessary groundwork for further optimization of limit, such as limit pushdown or whole-stage codegen.

At a high-level, this replaces the old physical `Limit` operator with two new operators, `LocalLimit` and `GlobalLimit`. `LocalLimit` performs per-partition limits, while `GlobalLimit` applies the final limit to a single partition; `GlobalLimit`'s declares that its `requiredInputDistribution` is `SinglePartition`, which will cause the planner to use an `Exchange` to perform the appropriate shuffles. Thus, a logical `Limit` appearing in the middle of a query plan will be expanded into `LocalLimit -> Exchange to one partition -> GlobalLimit`.

In the old code, calling `someDataFrame.limit(100).collect()` or `someDataFrame.take(100)` would actually skip the shuffle and use a fast-path which used `executeTake()` in order to avoid computing all partitions in case only a small number of rows were requested. This patch preserves this optimization by treating logical `Limit` operators specially when they appear as the terminal operator in a query plan: if a `Limit` is the final operator, then we will plan a special `CollectLimit` physical operator which implements the old `take()`-based logic.

In order to be able to match on operators only at the root of the query plan, this patch introduces a special `ReturnAnswer` logical operator which functions similar to `BroadcastHint`: this dummy operator is inserted at the root of the optimized logical plan before invoking the physical planner, allowing the planner to pattern-match on it.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Ping @rxin, since this is addressing an old TODO that you added.

**Comment by JoshRosen** [comment]
One question that I have: will the change here introduce a performance regression in cases where Limit is the final operator in the plan?  I think that we may want to continue to do the `executeTake()` for those cases, which may require us to use a different physical plan for `Limit` in cases where it's the terminal operator.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala [hunk 1]
```diff
@@ -108,48 +105,24 @@ case class Union(children: Seq[SparkPlan]) extends SparkPlan {
...
-  // TODO: Implement a partition local limit, and use a strategy to generate the proper limit plan:
-  // partition local limit -> exchange into one partition -> partition local limit again
-
-  /** We must copy rows when sort based shuffle is on */
-  private def sortBasedShuffleOn = SparkEnv.get.shuffleManager.isInstanceOf[SortShuffleManager]
-
   override def output: Seq[Attribute] = child.output
-  override def outputPartitioning: Partitioning = SinglePartition
 
   override def executeCollect(): Array[Row] = child.executeTake(limit)
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: Whoops, forgot to remove this.  This should probably be split to a separate operator in order to handle the cases where Limit is the final operator.

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: I looked at the code for `executeTake` and it looks like it's also doing some potentially-unnecessary copying and inefficient serialization of result rows (AFAIK it's not using SqlSerializer2).

**Comment by JoshRosen** [comment]
I think that this patch should be revisited after refactoring `take()` and `collect()` to use the efficient SQL serializers.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
(Testing a pull request builder configuration change...)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 2]
```diff
@@ -332,8 +332,12 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
         execution.Sample(lb, ub, withReplacement, seed, planLater(child)) :: Nil
       case logical.LocalRelation(output, data) =>
         LocalTableScan(output, data) :: Nil
-      case logical.Limit(IntegerLiteral(limit), child) =>
-        execution.Limit(limit, planLater(child)) :: Nil
+      case logical.Limit(IntegerLiteral(limit), child) => {
```
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Nit: Unnecessary `{  }`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 3]
```diff
@@ -332,8 +332,12 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
...
+        val perPartitionLimit = execution.PartitionLocalLimit(limit, planLater(child))
+        val globalLimit = execution.PartitionLocalLimit(
+          limit, execution.Exchange(SinglePartition, Nil, perPartitionLimit))
```
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Instead of manually specify  a single `SinglePartition` in `SparkStrategies`, a better idea probably specifying the `requiredDistribution` in the physical operator `execution.PartitionLocalLimit`, as we did in `execution.Aggregate`, the rational we do that is, in the case of the `perPartitionLimit` is a `SinglePartition` already, we don't need to do anther shuffle, the `EnsureRequirements` can handle such kind of optimization for us.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala (hunk 1 above)
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: Yes, this will call the `submitJob`, probably we don't have opportunity to use the `SqlSerializer2`, as it will be done by `SparkEnv.serializer` eventually, is I understand correctly.
At least we'd better do it in a separate PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala (hunk 3 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Good idea.  I think we'll have to add something akin to `partial=True` so that we can handle the per-partition and global limits properly.  I'll push another commit shortly to explore this option.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala (hunk 1 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala]: I agree.  I was planning to submit separate JIRA + pull request to address this for take/collect.  I originally thought that we might want that pull request to go in ahead of this one but I now think that they're fairly distinct and don't need to block each other.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala (hunk 3 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Actually, `partial` might be confusing so I'll just have two separate operators, `PartitionLocalLimit` and `GlobalLimit`.

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: _Actually_, I take _that_ back: maybe `global` is the right name for the parameter, similar to Sort.

**Comment by JoshRosen** [comment]
Alright, I've updated this to allow the planner to insert the Exchange.  We still can't merge this as-is, though, since it will introduce a bad regression for the case where Limit is the final operator in a physical plan.  AFAIK there's no easy way to write optimizer rules which reference the parent operator, so it's hard to write an optimizer rule to recognize this pattern and use a separate Limit implementation.

**Comment by chenghao-intel** [comment]
@JoshRosen ,seems `execution.CollectLimit` will eventually invoke the code like (in SparkPlan.executeTake):
```scala
sc.runJob(childRDD, (it: Iterator[InternalRow]) => it.take(left).toArray, 
  p,  allowLocal = false)
```
I am wondering if 
`execution.CollectLimit(limit, planLater(child))` 
V.S. 
`execution.Limit(global = true, limit, execution.Limit(global=false, limit, child))`
are actually equals in data shuffling / copying, if so, probably we can simplify the code by removing the `CollectLimit` and `ReturnAnswer`. 

Sorry if I missed something.

**Comment by JoshRosen** [comment]
@chenghao-intel:

They are not equivalent in terms of data shuffling, which is why I introduced `CollectLimit`.  In the old code, calling `executeCollect` on a `Limit` operator would end up running a code path which skips the shuffle after the per-partition limits and uses `executeTake` to try to only compute as many partitions as are necessary to answer the query.  `execution.Limit(global = true, limit, execution.Limit(global=false, limit, child))`, on the other hand, will end up involving a shuffle, so we'll end up calling `executeTake` on an RDD with a single partition, preventing it from being able to do less work.

**Comment by JoshRosen** [comment]
There's one problem with the planner changes here: during logical + physical planning we don't know in advance whether `executeTake()` is going to be called.  If I was to call `.rdd` on a query plan with a limit as the final operator then I would get an error.

**Comment by JoshRosen** [comment]
One hacky approach which would not require any planner changes would be to change `Limit` to hold a reference to the original child being limited.  If `executeCollect()` is called on the post-shuffle `Limit`, then it could just fall back to calling `take()` on the pre-shuffle RDD.

**Comment by JoshRosen** [comment]
Closing for now, given fact that limit should generally not appear in the middle of query plans unless something has gone horribly wrong.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by gatorsmile** [comment]
retest it please

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 4]
```diff
@@ -44,7 +44,7 @@ class QueryExecution(val sqlContext: SQLContext, val logical: LogicalPlan) {
 
   lazy val sparkPlan: SparkPlan = {
     SQLContext.setActive(sqlContext)
-    sqlContext.planner.plan(optimizedPlan).next()
+    sqlContext.planner.plan(ReturnAnswer(optimizedPlan)).next()
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: cc @marmbrus here to make sure it is safe

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: Yeah, this looks good to me.

**Comment by rxin** [comment]
cc @davies for a more detailed review

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+ * Take the first `limit` elements and collect them to a single partition.
+ *
+ * This operator will be used when a logical `Limit` operation is the final operator in an
+ * logical plan, which happens when the user is collecting results back to the driver.
+ */
+case class CollectLimit(limit: Int, child: SparkPlan) extends UnaryNode {
+  override def output: Seq[Attribute] = child.output
+  override def outputPartitioning: Partitioning = SinglePartition
+  override def executeCollect(): Array[InternalRow] = child.executeTake(limit)
+  protected override def doExecute(): RDD[InternalRow] = sparkContext.makeRDD(executeCollect(), 1)
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: This should not be called, should we throw UnsupportedOperaionError?

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: This actually _can_ get called if you call `df.limit(10).cache()`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 6]
```diff
@@ -337,8 +337,12 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
         execution.Sample(lb, ub, withReplacement, seed, planLater(child)) :: Nil
       case logical.LocalRelation(output, data) =>
         LocalTableScan(output, data) :: Nil
+      case logical.ReturnAnswer(logical.Limit(IntegerLiteral(limit), child)) =>
+        execution.CollectLimit(limit, planLater(child)) :: Nil
       case logical.Limit(IntegerLiteral(limit), child) =>
-        execution.Limit(limit, planLater(child)) :: Nil
+        val perPartitionLimit = execution.LocalLimit(limit, planLater(child))
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Should we have a test for this?

```
   select * from (select * from A limit 10 ) t1 UNION ALL (select * from B limit 10) t2
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 5 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: For this case, should we use GlobalLimit?

Or it will waste some time in collecting rows to driver, than sending them to executors,

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala (hunk 6 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: Today this will be planned as a `scan -> local limit -> global limit` for each branch of the union, which I believe is correct: this query should return at most 20 results.

What do you think this test should be asserting?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: I'm thinking that we did not have this kind of test case, right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 5 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: This is actually a tricky problem to address. If we want to actually plan `GlobalLimit` here instead then we need to be able to distinguish between the collect-to-driver and cache cases earlier in the planning process, which I think is going to be a large and non-trivial change.

If we want to avoid collecting back to the driver and re-broadcasting then I suppose we can do something similar to what the old `Limit` operator's `doExecute` did. The problem with the old `doExecute()` was that it involved a lot of inefficient copying and serialization. Instead, it might be worth seeing whether the inner logic of the `Exchange` operator could be called from `CollectLimit.doExecute`.

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: Similar to ReturnAnswer, we could put a CachePlan() as the root (See CacheManager.cacheQuery), and have a special case for it.

If it's not trivial, we could just leave a TODO for this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala (hunk 6 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: This is not allowed. See my PR: https://github.com/apache/spark/pull/10689

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: @davies, to clarify, are you proposing that we need to add a test to ensure that global limits below union are not pulled up?

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: @gatorsmile: I think we're in agreement. To recap:

Before this patch, `   select * from (select * from A limit 10 ) t1 UNION ALL (select * from B limit 10) t2` should get planned as:

```
          ┌─────────┐         
          │  Union  │         
          └─────────┘         
               ▲              
      ┌────────┴────────┐     
      │                 │     
┌──────────┐      ┌──────────┐
│ Limit 10 │      │ Limit 10 │
└──────────┘      └──────────┘
      ▲                 ▲     
      │                 │     
┌──────────┐      ┌──────────┐
│  Scan A  │      │  Scan B  │
└──────────┘      └──────────┘
```

Afterwards, this becomes

```
            ┌─────────┐             
            │  Union  │             
            └─────────┘             
                 ▲                  
        ┌────────┴──────────┐       
        │                   │       
┌───────────────┐   ┌──────────────┐
│GlobalLimit 10 │   │GlobalLimit 10│
└───────────────┘   └──────────────┘
        ▲                   ▲       
        │                   │       
┌──────────────┐    ┌──────────────┐
│LocalLimit 10 │    │ LocaLimit 10 │
└──────────────┘    └──────────────┘
        ▲                   ▲       
        │                   │       
  ┌──────────┐        ┌──────────┐  
  │  Scan A  │        │  Scan B  │  
  └──────────┘        └──────────┘  
```

What is **not** legal to do here is to pull the `GlobalLimit` up, so the following would be wrong:

```
         ┌───────────────┐          
         │GlobalLimit 10 │          
         └───────────────┘          
                 ▲                  
                 │                  
            ┌─────────┐             
            │  Union  │             
            └─────────┘             
                 ▲                  
        ┌────────┴──────────┐       
        │                   │       
┌──────────────┐    ┌──────────────┐
│LocalLimit 10 │    │ LocaLimit 10 │
└──────────────┘    └──────────────┘
        ▲                   ▲       
        │                   │       
  ┌──────────┐        ┌──────────┐  
  │  Scan A  │        │  Scan B  │  
  └──────────┘        └──────────┘  
```

That plan would be semantically equivalent to executing

```
  select * from (select * from A ) t1 UNION ALL (select * from B) t2 LIMIT 10
```

@davies, were you suggesting that the current planning is wrong? Or that we need more tests to guard against incorrect changes to limit planning? I don't believe that the changes in this patch will affect the planning of the case being described here, since we're not making any changes to limit pull-up or push-down. I _do_ have a followup patch in the works which takes @gatorsmile's two limit-pushdown patches and rebases them on top of the changes here: https://github.com/apache/spark/compare/master...JoshRosen:limit-pushdown-2. In that patch, I do plan to add more tests to handle these pushdown-related concerns.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 5 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: That's a really good idea. I think we could just match on `CachePlan(CollectLimit(...))` and replace it by `GlobalLimit`. Let me go ahead and try this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala (hunk 6 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala]: @JoshRosen Thank you for your explanation! That is so great that my previous PR are useful. So far, I am unable to find more operators for limit push down, except outer join and union all:  https://github.com/apache/spark/pull/10454 and https://github.com/apache/spark/pull/10451

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 5 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: I've started working on this and ran into one minor snag: what happens if a user has a DataFrame whose root logical plan is an RDD, then calls `.rdd()` on it? In that case, `CollectLimit` will still be the root and we'll need it to have a functioning `doExecute()`.

On reflection, I agree that the current `doExecute()` here is bad because it breaks linage and laziness. I'll see about switching back to an approach which invokes Exchange directly.

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: Maybe having a special case for Limit as root is not a good idea, I think we could always have GlobalLimit, call `executeTake(limit)` in that.

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: As discussed upthread, simply always having `GlobalLimit` would lead to regressions when calling `df.limit(10).collect()`, since it would now have to compute all partitions of `df` whereas before it might be able to compute only a subset.

**Comment by JoshRosen** [comment]
Alright, just pushed an update which I think should fix the laziness problems with `.rdd` and `.cache`. The key idea is to move the `ReturnAnswer` injection into `DataFrame.collect()` rather than always putting it at the root of the logical plan before invoking the planner.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by davies** [comment]
LGTM, pending tests.

**Comment by JoshRosen** [comment]
This should now be ready for a final sign-off.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+ * logical plan, which happens when the user is collecting results back to the driver.
+ */
+case class CollectLimit(limit: Int, child: SparkPlan) extends UnaryNode {
+  override def output: Seq[Attribute] = child.output
+  override def outputPartitioning: Partitioning = SinglePartition
+  override def executeCollect(): Array[InternalRow] = child.executeTake(limit)
+  private val serializer: Serializer = new UnsafeRowSerializer(child.output.size)
+  protected override def doExecute(): RDD[InternalRow] = {
+    val shuffled = new ShuffledRowRDD(
+      Exchange.prepareShuffleDependency(child.execute(), child.output, SinglePartition, serializer))
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: In this case, the plan in SparkUI (no exchange) will not match the actual plan (having exchange)

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: Since the current implementation also adds an ShuffledRowRDD without Exchange, this is not a regression.

**Comment by davies** [comment]
LGTM, merging this into master.
